### ISSUE TITLE
Investigate Pytorch AutoML is 4.6% worse than TF AutoML for MercedesBenzGreener

### ISSUE THREAD

**Issue by amholler** [ludwig-ai/ludwig#1697]
Investigate Pytorch AutoML 4.6% worse than TF AutoML for MercedesBenzGreener
```
Background: The MercedesBenzGreener dataset was one of the 12 datasets used to form the AutoML heuristics.

For the MercedesBenzGreener dataset, RMSE for AutoML running for 1 hr using pytorch/master (8.437) is 4.6%
worse than RMSE running for 1 hr using tf/tf-legacy (8.051).  This ticket tracks investigating that difference.
 https://docs.google.com/spreadsheets/d/1c1ghzlNBGH8Sh0AgxZw8Z366NZegiblxOuwQB9eHHf4/edit#gid=2057342027

This script can be used to run autoML for 1 hour:
 https://github.com/ludwig-ai/experiments/blob/main/automl/heuristics/mercedes_benz_greener/run_auto_train_1hr.py
The run will produce hyperopt_statistics.json.

This script can be used to post-process the resulting hyperopt_statistics.json to get the test RMSE:
 https://github.com/ludwig-ai/experiments/blob/main/utils/best_hyperopt_statistics.py
For example, if you are in the directory experiments/automl/heuristics/mercedes_benz_greener, run
 python ../../../utils/best_hyperopt_statistics.py hyperopt_statistics.json root_mean_squared_error

Note that MercedesBenzGreener is a small wide dataset and that tf itself is showing signs of overfit in the sanity
check runs of tf AutoML against the full heuristics runs; the validation split is better than the baseline run from the
heuristics search but the test split is performing worse.
 https://docs.google.com/spreadsheets/d/1c1ghzlNBGH8Sh0AgxZw8Z366NZegiblxOuwQB9eHHf4/edit#gid=475619705
 This is mentioned in the Ludwig AutoML blog draft
  https://docs.google.com/document/d/1AF1gDGqqaQkNcMKEgrNHgxB275ZBemdta5lk1SBLdxM/edit#
  ```

**Comment by dantreiman** [comment]
I can consistently repro this, I get [not exactly the same] but similar numbers:
```
(ludwig39-dev) daniel@Daniel-M1-Pro mercedes_benz_greener % python ../../../utils/best_hyperopt_statistics.py hyperopt_statistics.json root_mean_squared_error
best metric_score offset in validation 196
best metric_score hyperparameters {'combiner.bn_momentum': 0.8, 'combiner.bn_virtual_bs': 512, 'combiner.num_steps': 9, 'combiner.output_size': 16, 'combiner.relaxation_factor': 1.0, 'combiner.size': 24, 'combiner.sparsity': 0.001, 'training.batch_size': 256, 'training.decay_rate': 0.9, 'training.decay_steps': 8000, 'training.learning_rate': 0.02}
offset metric_score in test  74.94971466064453
offset metric_score in vali  60.113800048828125
offset metric_score in train 64.52440643310547
root_mean_squared_error offset in test  y 8.664525985717773
root_mean_squared_error offset in vali  y 7.690894603729248
root_mean_squared_error offset in train y 8.048173904418945
```

Out of curiosity, what are the best hyperparameters from the Tensorflow version?

**Comment by amholler** [comment]
>Out of curiosity, what are the best hyperparameters from the Tensorflow version?

Sorry, unfortunately, I didn't record this information.  It can be rerun on tf-legacy to find out.

**Comment by dantreiman** [comment]
Running latest tf-legacy I get:

```
(ludwig39-tf) daniel@Falcon3090:~/github/experiments/automl/heuristics/mercedes_benz_greener$ python ../../../utils/best_hyperopt_statistics.py hyperopt_statistics.json root_mean_squared_error
best metric_score offset in validation 73
best metric_score hyperparameters {
    'combiner.bn_momentum': 0.9,
    'combiner.bn_virtual_bs': 256,
    'combiner.num_steps': 5,
    'combiner.output_size': 64,
    'combiner.relaxation_factor': 1.5,
    'combiner.size': 64,
    'combiner.sparsity': 0.0001,
    'training.batch_size': 512,
    'training.decay_rate': 0.95,
    'training.decay_steps': 20000,
    'training.learning_rate': 0.005
}
offset metric_score in test  76.51455688476562
offset metric_score in vali  59.43547439575195
offset metric_score in train 67.27835845947266
root_mean_squared_error offset in test  y 8.949421882629395
root_mean_squared_error offset in vali  y 7.709438323974609
root_mean_squared_error offset in train y 8.217998504638672
```

**Comment by dantreiman** [comment]
Interesting, if I take the tensorflow-derived parameters and run them in PyTorch, I get:
```
Best validation model epoch: 55
Best validation model root_mean_squared_error on validation set y: 8.020873069763184
Best validation model root_mean_squared_error on test set y: 7.8994951248168945
```

**Comment by amholler** [comment]
I was running on a 3 node gpu-enabled ray cluster.
I was running ToT master [with blocker reverted] vs ToT tf-legacy as of 9 days ago.

**Comment by dantreiman** [comment]
PyTorch results
[trainable_func_fyIe89i.zip](https://github.com/ludwig-ai/ludwig/files/7997138/trainable_func_fyIe89i.zip)

**Comment by dantreiman** [comment]
Tensorflow Results (with checkpoints deleted to save space):
https://drive.google.com/file/d/1kiXceTY8tm5KHEw_X1-un8ZlWWqgFkLC/view?usp=sharing

**Comment by dantreiman** [comment]
PyTorch on CPU results:
[](https://drive.google.com/file/d/1h6hiHopm6FgpTYya_7ZMxHUIDQmD2Z8J/view?usp=sharing)

**Comment by dantreiman** [comment]
@anneholler Could you share the cluster config you are using to run on?  Thanks!  I keep getting different results on GPU, it would be helpful to know if I'm running on the same stack and hardware.

**Comment by dantreiman** [comment]
I got a good run on GPU using 3 g4.dn.4xlarge nodes.  Configs and metrics on this spreadsheet:
https://docs.google.com/spreadsheets/d/1e-pZDJj01F3irb0WEGWf5HwuLPcMgIQ9lwWz1ourBrk/edit?usp=sharing

Its hard to say exactly what is going on.  PyTorch best config is getting lower error on training set, but higher on test and val.  So the PyTorch best config is overfitting more.

Also, when comparing the learning curves (of the best respective training runs), I noticed an unusual spike in the PyTorch results.

![TF learning curve](https://user-images.githubusercontent.com/687280/153478332-2233c40d-055f-4cdf-8a30-efa2f0481de8.png)
![PT learning curve](https://user-images.githubusercontent.com/687280/153478440-79f8ef46-8833-4b00-8700-c94773ac656d.png)

Not sure yet what is causing this or if it is relevant, more investigation needed.

**Comment by amholler** [comment]
BTW, @dantreiman I notice that this dataset is also getting differences in type assignments between
tf-legacy and master (like we discussed yesterday from the memory point of view in KDD):
```
(tfenv) MacBook-Air-2:mercedes_benz_greener anne$ diff auto_config.json.automl auto_config.json.automl.master
84,86c84,86
<  'input_features': [{'column': 'X0', 'name': 'X0', 'type': 'category'},
<                     {'column': 'X1', 'name': 'X1', 'type': 'category'},
<                     {'column': 'X2', 'name': 'X2', 'type': 'category'},
---
>  'input_features': [{'column': 'X0', 'name': 'X0', 'type': 'text'},
>                     {'column': 'X1', 'name': 'X1', 'type': 'text'},
>                     {'column': 'X2', 'name': 'X2', 'type': 'text'},
89c89
<                     {'column': 'X5', 'name': 'X5', 'type': 'category'},
---
>                     {'column': 'X5', 'name': 'X5', 'type': 'text'},
91c91
<                     {'column': 'X8', 'name': 'X8', 'type': 'category'},
---
>                     {'column': 'X8', 'name': 'X8', 'type': 'text'},
467d466
<               'regularization_lambda': 1,
(tfenv) MacBook-Air-2:mercedes_benz_greener anne$ pwd
/Users/anne/ludwig/experiments/automl/heuristics/mercedes_benz_greener
```
I don't know if this could cause differences in model performance.

**Comment by amholler** [comment]
Note that the 'regularization_lambda': 1, difference is expected.

**Comment by dantreiman** [comment]
@amholler I think you are right about the input types.  I ran this again, using the `user_config` option to force both versions to use the same input data types.

Good news: both versions land on mostly similar parameters, same batch size, and closer final metric score 60.0685 (tf) vs. 60.6626 (torch)

Code:
```
with open('user_config.yaml', 'rb') as f:
    user_config = yaml.safe_load(f)

mercedes_benz_greener_df = load_mercedes_benz_greener()

auto_train_results = auto_train(
    dataset=mercedes_benz_greener_df,
    target='y',
    time_limit_s=3600,
    tune_for_memory=False,
    user_config=user_config
)
```

Pytorch results:
```
+----------------+------------+-------------------+------------------------+--------------------------+----------------------+------------------------+------------------------------+-----------------+---------------------+----------------------+----------------------+-----------------------+-------------------------+--------+------------------+----------------+
| Trial name     | status     | loc               |   combiner.bn_momentum |   combiner.bn_virtual_bs |   combiner.num_steps |   combiner.output_size |   combiner.relaxation_factor |   combiner.size |   combiner.sparsity |   trainer.batch_size |   trainer.decay_rate |   trainer.decay_steps |   trainer.learning_rate |   iter |   total time (s) |   metric_score |
|----------------+------------+-------------------+------------------------+--------------------------+----------------------+------------------------+------------------------------+-----------------+---------------------+----------------------+----------------------+-----------------------+-------------------------+--------+------------------+----------------|
| trial_dd97affc | TERMINATED | 172.31.9.92:436   |                   0.7  |                     2048 |                    4 |                      8 |                          1   |              32 |              0.0001 |                 8192 |                 0.95 |                   500 |                   0.025 |   1740 |        3596      |        64.4081 |
| trial_de0e35c8 | TERMINATED | 172.31.14.97:1457 |                   0.8  |                      256 |                    9 |                    128 |                          1   |              32 |              0      |                 2048 |                 0.95 |                 10000 |                   0.005 |     23 |          75.6883 |       341.896  |
| trial_de201716 | TERMINATED | 172.31.2.131:447  |                   0.98 |                     1024 |                    3 |                      8 |                          1   |               8 |              1e-06  |                 1024 |                 0.8  |                   500 |                   0.005 |     41 |          73.0759 |        91.5639 |
| trial_de909d10 | TERMINATED | 172.31.2.131:536  |                   0.8  |                     1024 |                    7 |                     24 |                          2   |               8 |              0.001  |                  256 |                 0.95 |                 20000 |                   0.01  |     23 |          74.1498 |       131.794  |
| trial_0c1caeb8 | TERMINATED | 172.31.14.97:2986 |                   0.9  |                      256 |                    5 |                     64 |                          1.5 |              64 |              0.0001 |                  512 |                 0.95 |                 20000 |                   0.005 |   1542 |        3514.13   |        63.3714 |
| trial_0ca851e8 | TERMINATED | 172.31.2.131:605  |                   0.6  |                      256 |                    3 |                     16 |                          1.2 |              16 |              0.01   |                 1024 |                 0.9  |                  8000 |                   0.005 |    215 |         361.174  |        69.0541 |
| trial_3ac218de | TERMINATED | 172.31.2.131:656  |                   0.6  |                      512 |                    5 |                     32 |                          2   |               8 |              0      |                 2048 |                 0.95 |                 10000 |                   0.025 |     39 |          72.9415 |       109.335  |
| trial_146bfa64 | TERMINATED | 172.31.2.131:707  |                   0.9  |                     4096 |                    7 |                     24 |                          1.5 |              64 |              0      |                  256 |                 0.9  |                 10000 |                   0.01  |     18 |          74.5411 |       121.671  |
| trial_424b2c16 | TERMINATED | 172.31.2.131:758  |                   0.8  |                      256 |                    3 |                     16 |                          1.2 |              64 |              0.001  |                 4096 |                 0.95 |                  2000 |                   0.005 |     32 |          72.1142 |      3221.51   |
| trial_71a724f6 | TERMINATED | 172.31.2.131:809  |                   0.8  |                      512 |                    9 |                     16 |                          1   |              24 |              0.001  |                  256 |                 0.9  |                  8000 |                   0.02  |    793 |        2838.42   |        60.6626 |
+----------------+------------+-------------------+------------------------+--------------------------+----------------------+------------------------+------------------------------+-----------------+---------------------+----------------------+----------------------+-----------------------+-------------------------+--------+------------------+----------------+
```

tf-legacy results:
```
+----------------+------------+-------------------+------------------------+--------------------------+----------------------+------------------------+------------------------------+-----------------+---------------------+-----------------------+-----------------------+------------------------+--------------------------+--------+------------------+----------------+
| Trial name     | status     | loc               |   combiner.bn_momentum |   combiner.bn_virtual_bs |   combiner.num_steps |   combiner.output_size |   combiner.relaxation_factor |   combiner.size |   combiner.sparsity |   training.batch_size |   training.decay_rate |   training.decay_steps |   training.learning_rate |   iter |   total time (s) |   metric_score |
|----------------+------------+-------------------+------------------------+--------------------------+----------------------+------------------------+------------------------------+-----------------+---------------------+-----------------------+-----------------------+------------------------+--------------------------+--------+------------------+----------------|
| trial_90f5e990 | TERMINATED | 172.31.1.26:1327  |                   0.7  |                     2048 |                    4 |                      8 |                          1   |              32 |              0.0001 |                  8192 |                  0.95 |                    500 |                    0.025 |   2609 |        3596.22   |        61.6253 |
| trial_9178f98e | TERMINATED | 172.31.15.14:408  |                   0.8  |                      256 |                    9 |                    128 |                          1   |              32 |              0      |                  2048 |                  0.95 |                  10000 |                    0.005 |      1 |         106.677  |     10221.2    |
| trial_ea90e446 | TERMINATED | 172.31.7.225:412  |                   0.98 |                     1024 |                    3 |                      8 |                          1   |               8 |              1e-06  |                  1024 |                  0.8  |                    500 |                    0.005 |     20 |          72.9318 |      8199.69   |
| trial_f5313b62 | TERMINATED | 172.31.7.225:633  |                   0.8  |                     1024 |                    7 |                     24 |                          2   |               8 |              0.001  |                   256 |                  0.95 |                  20000 |                    0.01  |     95 |         362.706  |        63.6851 |
| trial_24aa17a6 | TERMINATED | 172.31.15.14:558  |                   0.9  |                      256 |                    5 |                     64 |                          1.5 |              64 |              0.0001 |                   512 |                  0.95 |                  20000 |                    0.005 |      8 |          74.0691 |       453.44   |
| trial_30bd0454 | TERMINATED | 172.31.15.14:698  |                   0.6  |                      256 |                    3 |                     16 |                          1.2 |              16 |              0.01   |                  1024 |                  0.9  |                   8000 |                    0.005 |     15 |          73.7284 |      2247.54   |
| trial_61b1ddf0 | TERMINATED | 172.31.15.14:852  |                   0.6  |                      512 |                    5 |                     32 |                          2   |               8 |              0      |                  2048 |                  0.95 |                  10000 |                    0.025 |     11 |          74.2399 |       291.301  |
| trial_91d25ed8 | TERMINATED | 172.31.15.14:986  |                   0.9  |                     4096 |                    7 |                     24 |                          1.5 |              64 |              0      |                   256 |                  0.9  |                  10000 |                    0.01  |    783 |        3082.17   |        60.0685 |
| trial_c25540fc | TERMINATED | 172.31.7.225:1227 |                   0.8  |                      256 |                    3 |                     16 |                          1.2 |              64 |              0.001  |                  4096 |                  0.95 |                   2000 |                    0.005 |      8 |          72.9656 |      9512.51   |
| trial_015f5634 | TERMINATED | 172.31.7.225:1346 |                   0.8  |                      512 |                    9 |                     16 |                          1   |              24 |              0.001  |                   256 |                  0.9  |                   8000 |                    0.02  |      4 |          73.7548 |       247.073  |
+----------------+------------+-------------------+------------------------+--------------------------+----------------------+------------------------+------------------------------+-----------------+---------------------+-----------------------+-----------------------+------------------------+--------------------------+--------+------------------+----------------+
```

**Comment by dantreiman** [comment]
Learning curves are much more similar too, with input column types fixed:

![pytorch-master](https://user-images.githubusercontent.com/687280/154136431-cd24d65f-14db-485b-8a58-5f211467e37d.png)

![tf-legacy](https://user-images.githubusercontent.com/687280/154136454-ca6489a1-4e31-41b5-baa4-e62db56f0697.png)
